lib/Doctrine/ORM/AbstractQuery.php

```diff
@@ -529,9 +534,25 @@ private function translateNamespaces(Query\ResultSetMapping $rsm): void
      */
     public function setHydrationCacheProfile(?QueryCacheProfile $profile = null)
     {
-        if ($profile !== null && ! $profile->getResultCacheDriver()) {
-            $resultCacheDriver = $this->_em->getConfiguration()->getHydrationCacheImpl();
-            $profile           = $profile->setResultCacheDriver($resultCacheDriver);
+        if ($profile === null) {
+            $this->_hydrationCacheProfile = null;
+
+            return $this;
+        }
+
+        // DBAL < 3.2
+        if (! method_exists(QueryCacheProfile::class, 'setResultCache')) {
+            if (! $profile->getResultCacheDriver()) {
+                $defaultHydrationCacheImpl = $this->_em->getConfiguration()->getHydrationCacheImpl();
+                if ($defaultHydrationCacheImpl) {
+                    $profile = $profile->setResultCacheDriver($defaultHydrationCacheImpl);
+                }
+            }
+        } elseif (! $profile->getResultCache()) {
+            $defaultHydrationCacheImpl = $this->_em->getConfiguration()->getHydrationCacheImpl();
+            if ($defaultHydrationCacheImpl) {
+                $profile = $profile->setResultCache(CacheAdapter::wrap($defaultHydrationCacheImpl));
```

Comment:
Weird, the new tests I introduced do cover that path according to `var_dump();die();`

Reply:
Maybe https://github.com/doctrine/orm/pull/8964 did not take coverage into account?

Reply:
Working on it at https://github.com/doctrine/orm/pull/8998

Reply:
Looks better now.